File: packages/sdk/src/wrap.ts
import type {
  MCPToolCall,
  NodeExecution,
  NodeId,
  Policy,
  PolicyContext,
  SessionId,
  TraceEvent,
  TraceListener,
} from "@graphos-io/core";
import { PolicyViolationError } from "./errors.js";
import { extractMCPToolCalls } from "./policies/mcp-guard.js";

type StreamResult =
  | AsyncIterable<Record<string, unknown>>
  | Promise<AsyncIterable<Record<string, unknown>>>;

export interface GraphLike<TInput = unknown, TOutput = unknown> {
  invoke(input: TInput, config?: unknown): Promise<TOutput>;
  stream(input: TInput, config?: unknown): StreamResult;
}

export interface WrapOptions<TState = unknown> {
  policies?: Policy<TState>[];
  sessionId?: string;
  projectId?: string;
  onTrace?: TraceListener<TState>;
}

export interface WrappedGraph<TInput, TOutput> {
  invoke(input: TInput, config?: unknown): Promise<TOutput>;
  stream(
    input: TInput,
    config?: unknown
  ): AsyncIterable<Record<string, unknown>>;
}

const isPlainObject = (v: unknown): v is Record<string, unknown> =>
  v !== null && typeof v === "object" && !Array.isArray(v);

const mergeState = (
  acc: Record<string, unknown>,
  update: Record<string, unknown>
): Record<string, unknown> => {
  const out: Record<string, unknown> = { ...acc };
  for (const [k, v] of Object.entries(update)) {
    const prev = out[k];
    if (k === "messages") {
      const prevArr = Array.isArray(prev) ? prev : [];
      if (Array.isArray(v)) out[k] = [...prevArr, ...v];
      else if (v !== undefined && v !== null) out[k] = [...prevArr, v];
      else out[k] = prevArr;
    } else if (isPlainObject(prev) && isPlainObject(v)) {
      out[k] = mergeState(prev, v);
    } else {
      out[k] = v;
    }
  }
  return out;
};

let sessionCounter = 0;
const newSessionId = (): SessionId =>
  `gos_${Date.now().toString(36)}_${(sessionCounter++).toString(36)}` as SessionId;

const emit = <TState>(
  listener: TraceListener<TState> | undefined,
  event: TraceEvent<TState>
): void => {
  if (!listener) return;
  try {
    const result = listener(event);
    if (result && typeof (result as Promise<void>).then === "function") {
      (result as Promise<void>).catch((err) => {
        // eslint-disable-next-line no-console
        console.error("[graphos] onTrace listener rejected:", err);
      });
    }
  } catch (err) {
    // eslint-disable-next-line no-console
    console.error("[graphos] onTrace listener threw:", err);
  }
};

const emitMcpCalls = <TState>(
  listener: TraceListener<TState> | undefined,
  execution: NodeExecution<TState>,
  calls: MCPToolCall[]
): void => {
  for (const call of calls) {
    emit(listener, {
      kind: "mcp.call",
      sessionId: execution.sessionId,
      timestamp: execution.timestamp,
      step: execution.step,
      server: call.server,
      tool: call.tool,
      args: call.args,
      source: "graph",
    });
  }
};

export const GraphOS = {
  wrap<TInput, TOutput, TState = unknown>(
    graph: GraphLike<TInput, TOutput>,
    options: WrapOptions<TState> = {}
  ): WrappedGraph<TInput, TOutput> {
    const policies = options.policies ?? [];
    const { onTrace, projectId } = options;

    const runStream = async function* (
      input: TInput,
      config?: unknown
    ): AsyncIterable<Record<string, unknown>> {
      const sessionId = (options.sessionId ?? newSessionId()) as SessionId;
      const ctx: PolicyContext = { sessionId };
      for (const p of policies) p.reset?.(ctx);

      emit<TState>(onTrace, {
        kind: "session.start",
        sessionId,
        projectId,
        timestamp: Date.now(),
        input,
      });

      let step = 0;
      let lastStep = 0;

      const baseConfig =
        typeof config === "object" && config !== null
          ? (config as Record<string, unknown>)
          : {};
      const streamConfig =
        baseConfig.subgraphs === undefined
          ? { ...baseConfig, subgraphs: true }
          : baseConfig;

      try {
        const streamResult = graph.stream(input, streamConfig);
        const iterable = await Promise.resolve(streamResult);

        for await (const raw of iterable) {
          let path: readonly string[] = [];
          let chunk: Record<string, unknown>;
          if (
            Array.isArray(raw) &&
            raw.length === 2 &&
            Array.isArray((raw as unknown[])[0])
          ) {
            const tuple = raw as unknown as [string[], Record<string, unknown>];
            path = tuple[0];
            chunk = tuple[1];
          } else {
            chunk = raw as Record<string, unknown>;
          }

          const subgraphPrefix = path
            .map((seg) => seg.split(":")[0])
            .filter((seg): seg is string => !!seg)
            .join("/");

          for (const [node, state] of Object.entries(chunk)) {
            const qualifiedNode = subgraphPrefix
              ? (`${subgraphPrefix}/${node}` as NodeId)
              : (node as NodeId);
            const execution: NodeExecution<TState> = {
              sessionId,
              node: qualifiedNode,
              state: state as TState,
              step: step++,
              timestamp: Date.now(),
            };
            lastStep = execution.step;

            emit<TState>(onTrace, {
              kind: "step",
              sessionId,
              node: qualifiedNode,
              state: execution.state,
              step: execution.step,
              timestamp: execution.timestamp,
            });

            emitMcpCalls(onTrace, execution, extractMCPToolCalls(execution));

            for (const policy of policies) {
              const decision = policy.observe(execution, ctx);
              if (decision.kind === "halt") {
                emit<TState>(onTrace, {
                  kind: "policy.halt",
                  sessionId,
                  policy: decision.policy,
                  reason: decision.reason,
                  details: decision.details,
                  step: execution.step,
                  timestamp: Date.now(),
                });
                emit<TState>(onTrace, {
                  kind: "session.end",
                  sessionId,
                  timestamp: Date.now(),
                  outcome: "halted",
                });
                throw new PolicyViolationError(decision);
              }
            }
          }
          yield chunk;
        }

        emit<TState>(onTrace, {
          kind: "session.end",
          sessionId,
          timestamp: Date.now(),
          outcome: "complete",
        });
      } catch (err) {
        if (!(err instanceof PolicyViolationError)) {
          emit<TState>(onTrace, {
            kind: "session.end",
            sessionId,
            timestamp: Date.now(),
            outcome: "error",
            error: { message: err instanceof Error ? err.message : String(err) },
          });
        }
        throw err;
      }

      void lastStep;
    };

    return {
      stream: runStream,
      async invoke(input: TInput, config?: unknown): Promise<TOutput> {
        let merged: Record<string, unknown> = isPlainObject(input)
          ? { ...(input as Record<string, unknown>) }
          : {};
        for await (const chunk of runStream(input, config)) {
          for (const stateUpdate of Object.values(chunk)) {
            if (isPlainObject(stateUpdate)) {
              merged = mergeState(merged, stateUpdate);
            }
          }
        }
        return merged as TOutput;
      },
    };
  },
};

File: python/src/graphos_io/wrap.py
"""GraphOS.wrap — observe + guard a compiled LangGraph (Python).

The wrap is async-first because LangGraph Python's stream is ``astream()``.
We default to ``stream_mode="updates"`` and ``subgraphs=True`` so that
subgraph steps surface as qualified node names like ``response_agent/llm_call``
(matching the TypeScript SDK).
"""

from __future__ import annotations

import asyncio
import inspect
import logging
import time
from collections.abc import AsyncIterator, Iterable, Sequence
from typing import Any, Generic, Protocol, TypeGuard, TypeVar, runtime_checkable

from .errors import PolicyViolationError
from .policies.mcp_guard import extract_mcp_tool_calls
from .types import (
    MCPCallEvent,
    NodeExecution,
    NodeId,
    Policy,
    PolicyContext,
    PolicyHaltEvent,
    SessionEndEvent,
    SessionId,
    SessionStartEvent,
    StepEvent,
    TraceEvent,
    TraceListener,
    _Halt,
)

_LOGGER = logging.getLogger("graphos_io.wrap")

TInput = TypeVar("TInput")
TOutput = TypeVar("TOutput")
TState = TypeVar("TState")


@runtime_checkable
class CompiledGraph(Protocol):
    """Protocol matching LangGraph Python's compiled graph surface."""

    def astream(
        self, input: Any, config: Any | None = None, **kwargs: Any
    ) -> AsyncIterator[Any]: ...


def _now_ms() -> int:
    return int(time.time() * 1000)


_session_counter = 0


def _new_session_id() -> SessionId:
    global _session_counter
    sid = f"gos_{_now_ms():x}_{_session_counter:x}"
    _session_counter += 1
    return sid


def _is_plain_object(v: Any) -> TypeGuard[dict[str, Any]]:
    return isinstance(v, dict)


def _merge_state(
    acc: dict[str, Any], update: dict[str, Any]
) -> dict[str, Any]:
    """Generic state merger.

    Last-write-wins for scalars; concatenation for ``messages``; recursive
    merge for nested dicts. Matches LangGraph's default reducer plus
    ``add_messages`` for the most common shape.
    """

    out = dict(acc)
    for k, v in update.items():
        prev = out.get(k)
        if k == "messages":
            prev_arr: list[Any] = list(prev) if isinstance(prev, list) else []
            if isinstance(v, list):
                out[k] = prev_arr + list(v)
            elif v is not None:
                out[k] = [*prev_arr, v]
            else:
                out[k] = prev_arr
        elif _is_plain_object(prev) and _is_plain_object(v):
            out[k] = _merge_state(prev, v)
        else:
            out[k] = v
    return out


async def _emit(listener: TraceListener | None, event: TraceEvent) -> None:
    if listener is None:
        return
    try:
        result = listener(event)
        if inspect.isawaitable(result):
            await result
    except Exception:
        # Listener errors must never crash the wrapped graph.
        _LOGGER.warning("graphos_io: on_trace listener raised", exc_info=True)


def _normalize_chunk(raw: Any) -> tuple[Sequence[str], dict[str, Any]]:
    """Unpack a stream tuple ``(path, chunk)`` if subgraphs=True yielded one."""

    if (
        isinstance(raw, tuple)
        and len(raw) == 2
        and isinstance(raw[0], (list, tuple))
    ):
        path, chunk = raw
        if not isinstance(chunk, dict):
            return list(path), {}
        return list(path), chunk
    if isinstance(raw, dict):
        return [], raw
    return [], {}


def _qualified_node(prefix: Sequence[str], node: str) -> NodeId:
    if not prefix:
        return node
    cleaned = [seg.split(":", 1)[0] for seg in prefix if seg]
    if not cleaned:
        return node
    return f"{'/'.join(cleaned)}/{node}"


class WrappedGraph(Generic[TInput, TOutput]):
    """Returned by :func:`wrap`. Provides ``invoke`` and ``stream``."""

    def __init__(
        self,
        graph: CompiledGraph,
        *,
        policies: Sequence[Policy[Any]],
        session_id: SessionId | None,
        project_id: str | None,
        on_trace: TraceListener | None,
    ) -> None:
        self._graph = graph
        self._policies = list(policies)
        self._session_id = session_id
        self._project_id = project_id
        self._on_trace = on_trace

    async def invoke(self, input: TInput, config: Any | None = None) -> TOutput:
        """Run the graph to completion and return the merged final state."""

        merged: dict[str, Any] = dict(input) if _is_plain_object(input) else {}
        async for chunk in self.stream(input, config):
            for state_update in chunk.values():
                if _is_plain_object(state_update):
                    merged = _merge_state(merged, state_update)
        return merged  # type: ignore[return-value]

    async def stream(
        self, input: TInput, config: Any | None = None
    ) -> AsyncIterator[dict[str, Any]]:
        """Async-iterate the graph's node updates with policy enforcement."""

        session_id = self._session_id or _new_session_id()
        ctx = PolicyContext(sessionId=session_id)
        for policy in self._policies:
            reset = getattr(policy, "reset", None)
            if reset is not None:
                reset(ctx)

        await _emit(
            self._on_trace,
            SessionStartEvent(
                sessionId=session_id,
                projectId=self._project_id,
                timestamp=_now_ms(),
                input=input,
            ),
        )

        base_config: dict[str, Any] = dict(config) if _is_plain_object(config) else {}
        if "subgraphs" not in base_config:
            base_config["subgraphs"] = True
        if "stream_mode" not in base_config:
            base_config["stream_mode"] = "updates"

        step = 0
        try:
            iterable = self._graph.astream(input, base_config)
            async for raw in iterable:
                path, chunk = _normalize_chunk(raw)
                for node, state in chunk.items():
                    qualified = _qualified_node(path, node)
                    execution: NodeExecution[Any] = NodeExecution(
                        sessionId=session_id,
                        node=qualified,
                        state=state,
                        step=step,
                        timestamp=_now_ms(),
                    )
                    step += 1

                    await _emit(
                        self._on_trace,
                        StepEvent(
                            sessionId=session_id,
                            node=qualified,
                            state=state,
                            step=execution.step,
                            timestamp=execution.timestamp,
                        ),
                    )

                    for mcp_call in extract_mcp_tool_calls(execution):
                        await _emit(
                            self._on_trace,
                            MCPCallEvent(
                                sessionId=session_id,
                                timestamp=execution.timestamp,
                                step=execution.step,
                                server=mcp_call.server,
                                tool=mcp_call.tool,
                                args=mcp_call.args,
                                source="graph",
                            ),
                        )

                    for policy in self._policies:
                        decision = policy.observe(execution, ctx)
                        if isinstance(decision, _Halt):
                            await _emit(
                                self._on_trace,
                                PolicyHaltEvent(
                                    sessionId=session_id,
                                    policy=decision.policy,
                                    reason=decision.reason,
                                    details=decision.details,
                                    step=execution.step,
                                    timestamp=_now_ms(),
                                ),
                            )
                            await _emit(
                                self._on_trace,
                                SessionEndEvent(
                                    sessionId=session_id,
                                    timestamp=_now_ms(),
                                    outcome="halted",
                                ),
                            )
                            raise PolicyViolationError(decision)
                yield chunk

            await _emit(
                self._on_trace,
                SessionEndEvent(
                    sessionId=session_id,
                    timestamp=_now_ms(),
                    outcome="complete",
                ),
            )
        except PolicyViolationError:
            raise
        except (asyncio.CancelledError, KeyboardInterrupt):
            raise
        except Exception as exc:
            await _emit(
                self._on_trace,
                SessionEndEvent(
                    sessionId=session_id,
                    timestamp=_now_ms(),
                    outcome="error",
                    error={"message": str(exc)},
                ),
            )
            raise


def wrap(
    graph: CompiledGraph,
    *,
    policies: Iterable[Policy[Any]] | None = None,
    session_id: SessionId | None = None,
    project_id: str | None = None,
    on_trace: TraceListener | None = None,
) -> WrappedGraph[Any, Any]:
    """Wrap a compiled LangGraph with policies and telemetry.

    Returns a :class:`WrappedGraph` exposing ``invoke`` and ``stream``.
    """

    return WrappedGraph(
        graph,
        policies=list(policies) if policies else [],
        session_id=session_id,
        project_id=project_id,
        on_trace=on_trace,
    )


__all__ = ["CompiledGraph", "WrappedGraph", "wrap"]
